### PR TITLE
add possibility to mute yaml tests by operating system

### PR DESCRIPTION
**Pull request by hendrikmuhs**
this change adds the possibility to mute yaml tests based on operating system to avoid muting whole tests with the following syntax

```
- skip:
    feature: skip_os
    os: "OS/2 warp 4"
    reason: "rip"
```

**Important:** This uses the pretty name (full os name) reported by elasticsearch, not just `Linux` or `Windows`. The name should be accessible via the logs on CI (I will update this description after the 1st run, to show how it works.).

The extra `feature` specification ensures that the test does not break for clients that re-use the yaml tests. Clients parse the feature tag and use an inclusion list. As `skip_os` is not part of the inclusion list, clients will skip these tests.

~@henningandersen @droberts195 @danhermann WDYT?~

~I made this quickly in spare test triage cycles yesterday, because I have reservations in muting tests for _all_ systems due to the current memory accounting problems on debian 8.~

~The PR is not fully finished yet, but before spending more time on it, I like to have agreement that this is useful.~

### CODE REVIEW

**Comment by droberts195** [comment]
Thanks for doing this Hendrik.  I think this will be really useful functionality.

**Comment by henningandersen** [comment]
Yes, I also think this could be useful.

For the specific case in #66629, it might be faster though to fix it than to implement this mute capability? I.e., it might be a faster path to get rid of these errors in CI.

**Comment by nik9000** [comment]
Pinging clients team because they use those tests as well and they'd have to implement this in some way, though they might just implement it is "I always run my tests on windows/ubuntu or whatever so just skip if that is the string".

**Comment by hendrikmuhs** [comment]
> Pinging clients team because they use those tests as well and they'd have to implement this in some way, though they might just implement it is "I always run my tests on windows/ubuntu or whatever so just skip if that is the string".

I talked with the clients team, we agreed on adding a `skip_os` feature in addition, that way client parsers won't break:

```
- skip:
    feature: skip_os
    os: "OS/2 warp 4"
    reason: "rip"
```

We might remove this later, once _all_ clients added support for it. /CC @Mpdreamz

File: test/framework/src/main/java/org/elasticsearch/test/rest/yaml/section/SkipSection.java [hunk 1]
```diff
@@ -82,38 +86,52 @@ public static SkipSection parse(XContentParser parser) throws IOException {
...
         }
         if (Strings.hasLength(version) && Strings.hasLength(reason) == false) {
             throw new ParsingException(parser.getTokenLocation(), "reason is mandatory within skip version section");
         }
-        return new SkipSection(version, features, reason);
+        if (operatingSystems.isEmpty() == false && Strings.hasLength(reason) == false) {
+            throw new ParsingException(parser.getTokenLocation(), "reason is mandatory within skip version section");
+        }
+        // make feature "skip_os" mandatory if os is given, this is a temporary solution until language client tests know about os
+        if (operatingSystems.isEmpty() == false && features.contains("skip_os") == false) {
```
Comment (nik9000) [inline test/framework/src/main/java/org/elasticsearch/test/rest/yaml/section/SkipSection.java:109]: :+1:

**Review by nik9000 — APPROVED** [review]
LGTM

**Review by Mpdreamz — APPROVED** [review]

**Pull request merged**